fsql/deser.py

```diff
@@ -87,31 +91,51 @@ def read_and_concat(
 
 
 class PandasReader(DataReader):
+    """Wraps various pandas read methods (parquet, json, csv, excel) into a single interface.
+    Behaviour can be customised via passing any kwargs to the constructor.
+    """
+
+    def __init__(self, input_format=InputFormat.AUTO, **pdread_kwargs):
+        super().__init__(input_format=input_format)
+        self.pdread_user_kwargs = pdread_kwargs
+        self.pdread_default_kwargs = defaultdict(dict)
+        self.pdread_default_kwargs[InputFormat.PARQUET] = {
+            "engine": "fastparquet",
+        }
+        self.pdread_default_kwargs[InputFormat.JSON] = {
+            "lines": "true",
+        }
+        self.pdread_default_kwargs[InputFormat.XLSX] = {
+            "engine": "openpyxl",
+        }
+
     def read_single(self, partition: Partition, fs: AbstractFileSystem) -> pd.DataFrame:
         logger.debug(f"read dataframe for partition {partition}")
         input_format = self.detect_format(partition.url)
-        # TODO allow for user spec of engine and other params, essentially any quark
+        logger.debug(f"format detected for partition {input_format} <- {partition}")
         if input_format is InputFormat.PARQUET:
-            reader = lambda fd: pd.read_parquet(fd, engine="fastparquet")  # noqa: E731
+            reader = pd.read_parquet
         elif input_format is InputFormat.JSON:
-            reader = lambda fd: pd.read_json(fd, lines=True)  # noqa: E731
+            reader = pd.read_json
         elif input_format is InputFormat.CSV:
             reader = pd.read_csv
         elif input_format is InputFormat.XLSX:
-            reader = lambda fd: pd.read_excel(fd, engine="openpyxl")  # noqa: E731
+            reader = pd.read_excel
         elif input_format is InputFormat.AUTO:
             raise ValueError(f"partition had format detected as auto -> invalid state. Partition: {partition}")
         else:
             assert_exhaustive_enum(input_format)
 
+        pdread_kwargs = {**self.pdread_default_kwargs[input_format], **self.pdread_user_kwargs}
```

Comment:
I was thinking about some simple check that the given kwargs are valid for the selected reader method. But we can probably let pandas itself to fail and give a user reasonable error message.

Reply:
I'd probably rather fail... In some sense, our code is really a pass-through here -- so the less we do, the more risk of breaking/blocking something in the future... Philosophically, we are not trying to replace pandas reader or surpass it with a better interface, just to conveniently integrate it. Note that this check would require to work with all pandas versions, and we would have to be as informative as the original error cause -- so it may be just better to rely on it.